fractal_tasks_core/lib_zarr.py

```diff
@@ -0,0 +1,245 @@
+      sub-group;
+    * if the sub-group already exists and `overwrite=False`, fail.
+
+    Note that this version of the wrapper does not include the original
+    `dataset_kwargs` parameter.
+
+    Args:
+        group:
+            The group to write to.
+        key:
+            The key to write to in the group. Note that absolute paths will be
+            written from the root.
+        elem:
+            The element to write. Typically an in-memory object, e.g. an
+            AnnData, pandas dataframe, scipy sparse matrix, etc.
+
+    Raises:
+        OverwriteNotAllowedError:
+            If `overwrite=False` and the sub-group already exists.
+    """
+
+    # Set logger
+    if logger is None:
+        logger = logging.getLogger(None)
+
+    if key in set(group.group_keys()):
+        if not overwrite:
+            error_msg = (
+                f"Sub-group '{key}' of group {group.store.path} "
+                f"already exists, but `{overwrite=}`. "
+                "Hint: try setting `overwrite=True`."
+            )
+            logger.error(error_msg)
+            raise OverwriteNotAllowedError(error_msg)
+    write_elem(group, key, elem)
+
+
+def write_table(
+    image_group: zarr.Group,
+    table_name: str,
+    table: ad.AnnData,
+    overwrite: bool = False,
+    ngff_roi_attrs: Optional[dict[str, str]] = None,
+    logger: Optional[logging.Logger] = None,
+) -> zarr.group:
+    """
+    FIXME
+
+    Args:
+        image_group:
+            TBD
+        table_name:
+            TBD
+        table:
+            TBD
+        overwrite:
+            TBD
+        ngff_roi_attrs:
+            TBD
+        logger:
+            TBD
+
+    Returns:
+        Zarr group of the new table.
+    """
+
+    # Set logger
+    if logger is None:
+        logger = logging.getLogger(None)
+
+    # Create tables group, if needed
+    if "tables" not in set(image_group.group_keys()):
+        tables_group = image_group.create_group("tables", overwrite=False)
+    else:
+        tables_group = image_group["tables"]
+
+    # Check whether subgroup already exists, and proceed accordingly
+    if table_name in set(tables_group.group_keys()) and not overwrite:
+        raise
+
+    # If it's all OK, proceed and write the table
+    _write_elem_with_overwrite(
+        tables_group,
+        table_name,
+        table,
+        overwrite=overwrite,
+    )
+
+    # Update the `tables` metadata of the image group
+    current_tables = tables_group.attrs.asdict().get("tables") or []
+    if table_name in current_tables and not overwrite:
+        raise
+    if table_name not in current_tables:
+        new_tables = current_tables + [table_name]
+        tables_group.attrs["tables"] = new_tables
+
+    # Update OME-NGFF metadata for current-table group
+    # WARNING: the following OME-NGFF metadata are based on a proposed
+    # change to the specs (https://github.com/ome/ngff/pull/64)
+    if ngff_roi_attrs is not None:
```

Comment:
Could you document what the goal of this part is? Are we setting defaults for type & instance key with the code below?
And our ROI tables don't necessarily have a "label" obs column, right? e.g. the well_ROI_table isn't related to labels

Reply:
The goal is to (optionally) set the values of `ngff_table_attrs` (as per new naming).

In create-ome-zarr we will call `write_table` without the `ngff_table_attrs` parameter, as in
```python
write_table(
        image_group,
        output_ROI_table,
        bbox_table,
        overwrite=...,
    )
```
and then we won't enter this block.
This means that the `well_ROI_table` won't have any of the future ngff attributes (`type`, `region` and `instance_key`).

In some other tasks (namely when creating organoid-label-bounding-box ROIs via cellpose, that may be used as input ROIs for nuclear segmentation) we will call `write_table` as in 
```python
write_table(
        image_group,
        output_ROI_table,
        bbox_table,
        overwrite=...,
        ngff_table_attrs={
            "type": "ngff:region_table",
            "instance_key": "label",
            "region": {
                "path": f"../labels/{output_label_name}"
                },
            )
        )
```

Reply:
Hmm... It makes sense that a well_ROI_table wouldn't have an instance_key or region. They aren't following the PR 64 in NGFF, because ROI tables aren't the core logic the tables were proposed for.
I wouldn't say though that well_ROI_table should never have any metadata of this type. It may make sense if those tables had a type. I'm unsure whether "ngff:region_table" would be correct here. Sounds like a good fit for ROIs, but is meant more for those annotations specified in PR 64)

In conclusion: I think this function should eventually also assign some metadata to well_ROI_table and FOV_ROI_table. But given that it's not clear what that metadata should be, I'm fine with leaving it out until we can specify it better. Maybe as part of our discussions on formalizing our ROIs.

Reply:
Fine with me.
I started by abstracting the code we actually have, and I think we never include any other attribute apart from the `ngff:region_table`-related ones. It's easy to make this block more general if needed (we move to a general `table_attrs` parameter, and only if it includes the `ngff:region_table`-specific ones we perform additional checks that all the three are present).

Reply:
Good plan. Let's already switch it to the more general `table_attrs` now then, so we can easily add functionality to this later without having to change the function signature.

Reply:
> Good plan. Let's already switch it to the more general `table_attrs` now then, so we can easily add functionality to this later without having to change the function signature.

This is in principle done via https://github.com/fractal-analytics-platform/fractal-tasks-core/commit/76afceb0f297530fdeb286256057a08137ec9549, 6462017ed3adc8b0460bf0ecad6cd6d15c02b7ca and f85b3f8581a5c8b37a2056996feab3f454cd033b (see status of the PR at https://github.com/fractal-analytics-platform/fractal-tasks-core/pull/499/files).

Right now, I am just raising a warning if the proposed-specs attributes are not fully compliant, as in
```python
    # Optionally update attributes of the new-table zarr group
    if table_attrs is not None:
        if table_attrs.get("type") == "ngff:region_table":
            # Verify whether we comply with a proposed change to the OME-NGFF
            # table specs (https://github.com/ome/ngff/pull/64)
            try:
                table_attrs["instance_key"]
                table_attrs["region"]["path"]
            except KeyError as e:
                logger.warning(
                    f"The table_attrs parameter of write_elem has "
                    "type='ngff:region_table' but does not comply with the "
                    "proposed table specs. "
                    f"Original error: KeyError: {str(e)}"
                )
        # Overwrite all table_attrs key/value pairs into table_group attributes
        table_group.attrs.put(table_attrs)
```
Should this be an error or is it OK to keep a warning?

The question is relevant for the external use of the fractal-tasks-core library functions, while it's less likely that we get this wrong in our own tasks (if we test them).